### PR TITLE
Refactor rs-e2e project to remove Hapi HL7 library

### PR DESCRIPTION
**Pull request by basiliskus**
# Description

- Removed the `ca.uhn.hapi` dependencies
- Added `HL7Message`, `HL7Segment`, `HL7Encoding` and `HL7Path` classes to replace Hapi's `Message` class
- Improved and refactored HL7 parser and expression evaluator
- Improved handling of edge cases
- Reorganized file structure
- Added full test coverage

## Issue

#1532 
## Checklist

- [x] I have added tests to cover my changes
- [x] I have added JavaDocs where required
- [x] I have updated the documentation accordingly

### CODE REVIEW

**Comment by jbiskie** [comment]
Tested and looks good.

Sonar flagged a circular dependency between HL7Message and HL7Parser, and there are actually a lot of them. Should we consider tagging this a a separate task to clean up?

**Comment by basiliskus** [comment]
> Sonar flagged a circular dependency between HL7Message and HL7Parser, and there are actually a lot of them. Should we consider tagging this a a separate task to clean up?

I'll take a look. If it's straightforward to fix, we can include it in this PR

**Comment by jorg3lopez** [comment]
Test task fails when I run `./gradlew :rs-e2e:test --info`

image

**Comment by jorg3lopez** [comment]
> Test task fails when I run `./gradlew :rs-e2e:test --info`
> 
> image

I think we need to see the mocking for the rulesloader or rule. This is what is probably causing the null pointer:

image

image

**Comment by basiliskus** [comment]
@jorg3lopez I'm not able to replicate your error. We didn't make any changes to the assertion engine or its test in this PR. It's probably related to your local configuration. Can you try running the other the rule engine tests (ValidationRuleEngineTest and TransformationRuleEngineTest) and see if you get the same error?

**Comment by jorg3lopez** [comment]
> @jorg3lopez I'm not able to replicate your error. We didn't make any changes to the assertion engine or its test in this PR. It's probably related to your local configuration. Can you try running the other the rule engine tests (ValidationRuleEngineTest and TransformationRuleEngineTest) and see if you get the same error?

Per our conversation, this is a pre-existing bug that we will need to look into. This bug is present in the main repo, so ignore it for this PR.

File: rs-e2e/src/main/java/gov/hhs/cdc/trustedintermediary/rse2e/hl7/HL7Encoding.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+import java.util.Map;
+
+/** The HL7Encoding class represents the encoding characters used in an HL7 message. */
+public record HL7Encoding(Map<String, Character> characters) {
+    public static final String DEFAULT_SEGMENT_DELIMITER = "\n";
+    public static final char DEFAULT_FIELD_DELIMITER = '|';
+    public static final char DEFAULT_COMPONENT_DELIMITER = '^';
+    public static final char DEFAULT_REPETITION_DELIMITER = '~';
+    public static final char DEFAULT_ESCAPE_CHARACTER = '\\';
```
Comment (GilmoreA6) [inline rs-e2e/src/main/java/gov/hhs/cdc/trustedintermediary/rse2e/hl7/HL7Encoding.java:11]: Is there a reason the default escape character is not '\'. I'm assuming we need to escape the escape character for some reason?

Comment (basiliskus) [reply rs-e2e/src/main/java/gov/hhs/cdc/trustedintermediary/rse2e/hl7/HL7Encoding.java:11]: Right, we need to escape the escape character. That's because `\` is also an escape character in Java

File: rs-e2e/src/main/java/gov/hhs/cdc/trustedintermediary/rse2e/hl7/HL7Path.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    public static HL7Path parse(String path) {
+        Matcher matcher = HL7_FIELD_NAME_PATTERN.matcher(path);
+        if (!matcher.matches()) {
+            throw new HL7ParserException("Invalid HL7 path format: " + path);
+        }
+
+        String segmentName = matcher.group(1);
+        int[] indices =
+                Arrays.stream(matcher.group(2).split("\\.")).mapToInt(Integer::parseInt).toArray();
```
[the remark this answers is not in the document]
Comment (GilmoreA6) [reply rs-e2e/src/main/java/gov/hhs/cdc/trustedintermediary/rse2e/hl7/HL7Path.java:20]: I assume we don't need to catch this since the HL7_FIELD_NAME_PATTERN regex will already throw an error for invalid path format?

**Review by jbiskie — APPROVED** [review]

File: rs-e2e/src/main/java/gov/hhs/cdc/trustedintermediary/rse2e/hl7/HL7Path.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (basiliskus) [reply rs-e2e/src/main/java/gov/hhs/cdc/trustedintermediary/rse2e/hl7/HL7Path.java:20]: Right, the regex is already making sure the string are ints separated by dots. Otherwise it will throw an exception when matching

**Review by GilmoreA6 — APPROVED** [review]

File: rs-e2e/src/main/java/gov/hhs/cdc/trustedintermediary/rse2e/hl7/HL7Encoding.java (hunk 1 above)
Comment (somesylvie) [reply rs-e2e/src/main/java/gov/hhs/cdc/trustedintermediary/rse2e/hl7/HL7Encoding.java:11]: Will the `\n` also need a second `\` to escape its `\`? It would also be helpful to add a clarifying comment about the `\\`

Comment (basiliskus) [reply rs-e2e/src/main/java/gov/hhs/cdc/trustedintermediary/rse2e/hl7/HL7Encoding.java:11]: The `\` in `\n` is the escape character itself so we don't need another one. This is all standard Java syntax so not sure if we need to add further clarification. If you write a Windows path string in Java you would write `C:\\Program Files\\Java`, and if you write a newline in a string you'd write `Hello\nWorld`

Comment (somesylvie) [reply rs-e2e/src/main/java/gov/hhs/cdc/trustedintermediary/rse2e/hl7/HL7Encoding.java:11]: Ugh Java is so weird, thanks for the explanation
